futures-channel/src/mpsc/mod.rs

```diff
@@ -274,28 +282,19 @@ impl Error for TryRecvError {
 }
 
 #[derive(Debug)]
-struct Inner<T> {
-    // Max buffer size of the channel. If `None` then the channel is unbounded.
-    buffer: Option<usize>,
-
+struct UnboundedInner<T> {
     // Internal channel state. Consists of the number of messages stored in the
     // channel as well as a flag signalling that the channel is closed.
     state: AtomicUsize,
```

Comment:
This can be just an `AtomicBool` is_closed now, right?

Reply:
Unfortunately, it can't without changing the contract provided by the queue. Consider this scenario.

* sender checks `is_open` flag, it is open
* receiver calls `set_closed`
* receiver polls the queue, finds no messages, assumes the end of a stream (when there's no message counter)
* sender enqueues a message and returns the `Ok` result that the message is successfully sent
* so sender assumes the message is sent, and receiver already received the end of a stream, the message is lost

`message_count` field is needed to guarantee no message is lost during the queue closure. But it creates overhead.